File: Java/Medium/Maximum Width of Binary Tree.java


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 /*
    BFS solution. This solution goes through each level one by one. For each level, it
    records the position of the left-most node and the right-most node. Then, it 
    calculates the width of the level by subtracting those, and only records the longest
    width it finds.

    The position of each node is calculated like this:
        - left children get twice their parent's position
        - right children get twice their parent's position plus one

    Time complexity:
    O(N) because we visit every node once.

    Auxiliary space complexity:
    O(H), where H is the height of the tree. level_nodes stores 2 elements for every level,
    and the call stack reaches a size equivalent to the height at its peak.
    If the tree is balanced, we can say the SC is O(logN) because a balanced tree has a height
    of approximately log(base2)N.
    If it is skewed, we can say the SC is O(N) because the height may be linearly proportional
    to the number of nodes.
 */
class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        Queue<Object[]> q = new LinkedList<>();
        q.add(new Object[] {root, 0});
        int max_width = 0;

        // BFS
        while (q.size() > 0) {
            // At this point, exactly one entire level is stored in the queue
            // get the length so that we can traverse exactly this amount
            // also get the position of the first node in this level
            // and create a variable that will store the rightmost positions
            // as they are visited, so that the rightmost one is stored in the end
            int level_length = q.size();
            int leftmost = (int) q.peek()[1];
            int rightmost = leftmost;

            // Iterate through the current level
            for(int i = 0; i < level_length; i++) {
                Object[] curr_node_and_pos = q.poll();
                rightmost = (int) curr_node_and_pos[1];
                TreeNode curr_node = (TreeNode) curr_node_and_pos[0];
                int curr_position = (int) curr_node_and_pos[1];

                // Add children if not null
                if (curr_node.left != null)
                    q.add(new Object[] {curr_node.left, 2 * curr_position});

                if (curr_node.right != null)
                    q.add(new Object[] {curr_node.right, 2 * curr_position + 1});
            }

            // calculate width of the level we just traversed, keep it if it is a max
            // and move on to the next level
            max_width = Math.max(max_width, rightmost - leftmost + 1);
        }

        return max_width;
    }
}

File: TypeScript/Medium/Maximum Width of Binary Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

/* 
My better definition:
class TreeNode {
    constructor(
        public val: number = 0,
        public left: TreeNode | null = null,
        public right: TreeNode | null = null,
    ) {}
}
*/

function widthOfBinaryTree(root: TreeNode | null): number {
    let ans = 0;

    // BFS
    let queue = new Queue<[TreeNode, bigint]>();
    queue.push([root, 0n]);

    let curr: TreeNode;
    let currQueueNodes: number;
    // Use bigint because the indices multiply by 2 each level.
    // Since there can be 3000 nodes, there can be 3000 levels, so
    // pos can become 2^3000.
    let pos: bigint;
    let firstIdx: bigint;
    let lastIdx: bigint;
    while (!queue.isEmpty()) {
        // Process exactly the current layer
        currQueueNodes = queue.size();
        for (let i = 0; i < currQueueNodes; i++) {
            [curr, pos] = queue.pop();

            if (i === 0) firstIdx = pos;
            if (i === currQueueNodes - 1) lastIdx = pos;

            // Add neighbors
            if (curr.left)
                queue.push([curr.left, 2n * pos]);
            if (curr.right)
                queue.push([curr.right, 2n * pos + 1n]);
        }

        ans = Math.max(ans, Number(lastIdx - firstIdx + 1n))
    }

    return ans;
};

// Solved after seeing previous solutions
